### PR TITLE
Feature/#295 chexpert dataset

### PR DESCRIPTION
**Pull request by LenaZellinger**
I added the .py file and the Jupyter notebook for the preprocessing of CheXpert.
Please let me know if you have any suggestions for improvement.

Linked issue: #295

### CODE REVIEW

File: examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:105]: when you create a new path, it is always better to work with `os.path.join` (see https://docs.python.org/3/library/os.path.html#os.path.join), since e.g. "/" will not work for OSX/Linux but won't work for Windwos and vice versa

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:224]: subsection; mb "Image Preprocessing"?

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:225]:
```suggestion
    "[this](https://github.com/gaetandi/cheXpert/blob/master/cheXpert_final.ipynb) and [this](https://github.com/Stomper10/CheXpert/blob/master/CheXpert_DenseNet121_FL.ipynb)"
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:245]: one large paragraph is hard to read -> either bullet list, or blank lines.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:224]:
```suggestion
    "In the following, a transform sequence is defined, which will be used to preprocess the images. The images are first resized to 320 x 320 pixels and then normalized with the mean and standard deviation from ImageNet. Such transformations have previously been applied by several other submissions regarding CheXpert on GitHub such as:\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:247]:
```suggestion
    "The structure of the class was inspired by the GitHub submissions credited above."
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:258]: here and further: pls add typisation; an extensive guide is [here](https://www.python.org/dev/peps/pep-0484/)

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:256]:
```suggestion
    "class CheXpertDatasetProcessor(Dataset):\n",
```
Class names are always defined with camel-case

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:270]: normally no real work should be done in `def __init__()` but rather just the initialization of self parameters

then:
- in `def __init__()` there would be only lines 310-313
- the rest of the work should be split in different functions, so that 
     - one function that would be called in cell 11 (e.g. `def process_chexpert_dataset`)
     - other functions that would be called from it

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:310]: better to calculate it before and pass it as a function parameter

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:311]: same here

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:284]:
```suggestion
    "        if to_ones:\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:291]:
```suggestion
    "        if to_zeros:\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:298]:
```suggestion
    "        if to_ignore:\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:327]:
```suggestion
    "        if self.transform_sequence:\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:261]: Awesome that you added the docstrings!

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:266]:
```suggestion
    "        Args:\n"
    "            path: path to the folder where train.csv and valid.csv are stored\n",
    "            subset: either \"train\" to load the train.csv or \"valid\" to load valid.csv\n",
    "            to_ones: list of pathologies for which uncertainty labels should be replaced by 1\n",
    "            to_zeros: list of pathologies for which uncertainty labels should be replaced by 0\n",
    "            to_ignore: list of pathologies for which uncertainty labels should be ignored (label will be turned to nan)\n",
    "            transform_sequence: sequence used to transform the images\n",
    "        Returns:\n"
    " ADD WHAT IS RETURNED"
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:394]:
```suggestion
    "## Store the preprocessed data"
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:418]: the filenames should rather be defined at the begging of the tutorial (e.g. in the cell 2)

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:424]: same here

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:435]: make it as a final word in a separate subsection together with the references

**Review by anasedova — CHANGES_REQUESTED** [review]
Nice work! Here are a couple of my suggestions - pls have a look at them.

I would also ask you to mark the conversation as resolved when you made the correction and don't have any questions or add the comments in the corresponding conversation if smth is not clear/you do not agree with my suggestion/you have another idea - that would make it much easier for me to check everything afterward :)

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:42]: I'm not 100% sure whether I did the ordering correctly, I used the follwing page: https://docs.python.org/3/library/ to determine which imports are standard library imports and which are third party imports

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:146]: Yes, I saw the two follwing tables are even printed in the paper, so there shouldn't be a problem.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:418]: I will change that, once I have the section for storing the data finalized.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:58]: each IDE has its own hot keys to optimize the order of the imports, e.g. in PyCharm: https://www.jetbrains.com/help/pycharm/creating-and-optimizing-imports.html#optimize-imports you can check them

Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:58]: Thanks a lot, I'll try that!

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:682]: some (brief) explanation what all of that means would be good

Comment (anasedova) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:682]:
```suggestion
    "The next step is image transformation and preparation of the dataset for PyTorch's DataLoader, which is done with the `CheXpertDatasetProcessor` class. The uncertainty labels will be handled according to either the \"U-Ones\", \"U-Zeroes\", \"U-Ignore\" or \"U-MultiClass\" approach as they were described in the original CheXpert paper by Irvin et al. (2019).\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:684]: ah it is related to the previous methods, right? I see it now :D but I still would prefer a somewhat more consistent explanation. Maybe with a bullet list (four options, brief explanation + corresponding argument)?

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:688]:
```suggestion
    "If the `return_image` parameter is set to False, only the labels are returned. This might be useful if you already have the transformed images saved but wish to experiment with different settings of `to_ones`, `to_zeros`, or `to_ignore`.\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:701]: why does it inherit from the `Dataset` class?

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:708]:
```suggestion
    "                 transform_sequence: List = None,\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:756]: Oh I'm sorry, I have corrected it wrongly the last time. Ofc here and further it should be `if self.to_ones is not None` as you had had it before. I didn't get the logic of these parameters and thought they were booleans.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:833]:
```suggestion
    "For a given index, the `__getitem__` function returns the tensor representing the preprocessed image as well as a tensor containing all of the labels for this observation.\n",
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:835]:
```suggestion
    "Let's take a look at what `__getitem__` returns for the training example with index 0."
```

**Review by anasedova — CHANGES_REQUESTED** [review]
Awesome, basically only a bit of wording remains. Looking forward to the new data-saving part!

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:701]: I mainly chose that because most tutorials I looked up used it together with DataLoader (e.g.: https://pytorch.org/tutorials/beginner/basics/data_tutorial.html). Would something else be more suitable?

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:610]: Since Fracture only has label 0 (never 1) in the validation set, the value_counts function only returns a single number instead of 2 (as is the case with the other pathologies)

The output of np.array(validation_labels.value_counts(subset=["Fracture"]).sort_index(ascending=True)) is just array([234], dtype=int64), which means that Fracture has label 0 for 234 observations in the validation set (= all of the observations in the validation set). So val[1] in this case doesn't exist.

For all other pathologies we get 2 values as output, where the first number is the number of times the pathology is labeled as 0 and the second number is the number of times the pathology is labeled as 1, e.g. for Edema: array([189,  45], dtype=int64)

I will see if I can maybe find a function that automatically returns [234, 0] for Fracture, so that I don't need the exception in the loop.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb
Comment (anasedova) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:701]: Usually, if you inherit from some class, the "child" class shouldn't change the semantics of the "parent" class (e.g. if your parent class is Dataset (=the class that describes special datatype), you child class is not supposed to be the one where you *read* and *process* data). 

Also, the inheritance is usually needed when you want to reuse some arguments and/or functions from the parent class. Since you are not using any torch dataset properties here,  I don't think you need any inheritance here at all.

Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/Prepare_CheXpert_data.ipynb:701]: Thanks, that's good to know! I assumed that adding Dataset was necessary for DataLoader to work (since I mainly saw the 2 together) , but you're right, it seems to work without it, so I'll remove it.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py
Comment (AndSt) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:175]: I think it makes a lot of sense to write a Processor class, as you need a lot of information (labels) for each image. 
In general I think this is totally fine, but the functional style is more used in Big Data (cause easier to parallelize etc.).
Like:
```
1) df = transform_labels(df, **kwargs)
2) transform_images (path, df, **kwargs):
          for img in os.list(path):   # parallelizable
                  transform_img(path, img, df, **kwargs)
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py
Comment (AndSt) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:385]: More "pthonic"
```
for i, ex in enumerate(chexpert_train_resize): # with __getitem__ you can automatically loop through a structure:
      save_image(ex[0], os.path.join(storing_location, "train_images", "image_" + str(i) + ".jpg"))
```

File: examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py
Comment (AndSt) [inline examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:393]: In the notebook this function is run twice. Is this really needed?

**Review by AndSt — COMMENTED** [review]
Hello Lena. I have some minor things, in general really cool stuff! I looked at the Notebook but commented in the .py file, as it's easier ...
It's only a few style things.
Until the merge someone else should run the tutorial once. If I should do that please contact me, so we can figure out how to exchange dat.

File: examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py
Comment (LenaZellinger) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:62]: Hello Andy! Thanks a lot for looking at the files and for your suggestions! I will push the new files once I edited them according to your comments. 

It would be great if you could run the tutorial before the merge if you have time. I will contact you once the files are ready for that, thank you very much!

Comment (anasedova) [reply examples/data_preprocessing/[WIP]_chexpert_dataset/prepare_CheXpert_data.py:62]: @LenaZellinger @AndSt as for the tutorial - no problem, I can run it

**Comment by anasedova** [comment]
The pull request as it stands is merged to gradient matching repo as the data is not directly related to Knodel and is not compatible with Knodle.

**Pull request closed without merging**
